### PR TITLE
Add SubFeeFromAmount to options 

### LINKED ISSUE #386: Please automatically remember when the user checks 'Subtract fee from amount'
**Is your feature request related to a problem? Please describe.**
Yes. When sending transactions from bitcoin core, there is a checkbox to indicate whether the user would like to take the transaction fee from the amount they've entered. This decision isn't, however, remembered- and so funds can be assumed to be taken out of the rest of one's balance. This problem is very aggravating, as it undermines the privacy of using multiple receiving wallets to prevent traceability and mixing of one's fund history within the blockchain. 

**Describe the solution you'd like**
I would like by default for the transaction page to remember which settings you've chosen (especially the fee from amount checkbox), and for this default to be mutable in bitcoin core's options.

**Describe alternatives you've considered**
There aren't many alternatives other than tediously enabling this option every transaction after you didn't realize at first. I might not be aware of some cli method to achieve this, but if there were such a thing it would still be very user-unfriendly.

**Additional context**
image

### PR DESCRIPTION
**Pull request by psancheti110**
This PR adds **_SubFeeFromAmount_** option which lets the user select their preferred setting of whether fee for a transaction is to be subtracted from the amount or not for future transactions. The setting chosen by the user is remembered even when the GUI mode is turned off. 

**_Functionality and Usage:_** 

- Go to `Settings > Options > Wallet` on _Windows/Linux_ or `bitcoin-qt > Preferences > Wallet` on _macOS_.
- The checkbox **Subtract Fee From Amount** corresponds to the added option **SubFeeFromAmount**.
- The preferred setting intended to be the default for all future send transactions should be selected by the user.
- Click on **OK**.
- Go to the **Send** tab in the wallet.
- You shall notice, any new Send transaction created will have the preferred setting as chosen by the user.<br> (Try clicking on Add recipient or even restarting the Node in GUI)

Attaching ScreenRecordings to explain the added feature. 

> Master.mov: Master Branch

https://user-images.githubusercontent.com/54016434/127763378-be91837d-d0ab-4ae5-87c0-d303fa70a336.mov

> PR.mov: PullRequest

https://user-images.githubusercontent.com/54016434/127763404-05b834c1-4082-4fbd-9b05-1528ac898a21.mov

Close #386

### CODE REVIEW

**Comment by psancheti110** [comment]
This PR adds a **_SubFeeFromAmount_** option which lets the user select their preferred setting which is remembered every time they run the node on GUI mode.  

**Functionality :**

**Comment by psancheti110** [comment]
This PR adds **_SubFeeFromAmount_** option which lets the user select their preferred setting of whether fee for a transaction is to be subtracted from the amount or not for future transactions. The setting chosen by the user is remembered even when the GUI mode is turned off. 

**_Functionality and Usage:_** 

- Go to `Settings > Options > Wallet` on _Windows/Linux_ or `bitcoin-qt > Preferences > Wallet` on _macOS_.
- The checkbox **Subtract Fee From Amount** corresponds to the added option **SubFeeFromAmount**.
- The preferred setting intended to be the default for all future send transactions should be selected by the user.
- Click on **OK**.
- Go to the **Send** tab in the wallet.
- You shall notice, any new Send transaction created will have the preferred setting as chosen by the user.<br> (Try clicking on Add recipient or even restarting the Node in GUI)

Attaching ScreenRecordings to explain the added feature. 

> Master.mov: Master Branch

https://user-images.githubusercontent.com/54016434/127763378-be91837d-d0ab-4ae5-87c0-d303fa70a336.mov

> PR.mov: PullRequest

https://user-images.githubusercontent.com/54016434/127763404-05b834c1-4082-4fbd-9b05-1528ac898a21.mov


Close #386

**Comment by psancheti110** [comment]
This PR adds **_SubFeeFromAmount_** option which lets the user select their preferred setting of whether fee for a transaction is to be subtracted from the amount or not for future transactions. The setting chosen by the user is remembered even when the GUI mode is turned off. 

**_Functionality and Usage:_** 

- Go to `Settings > Options > Wallet` on _Windows/Linux_ or `bitcoin-qt > Preferences > Wallet` on _macOS_.
- The checkbox **Subtract Fee From Amount** corresponds to the added option **SubFeeFromAmount**.
- The preferred setting intended to be the default for all future send transactions should be selected by the user.
- Click on **OK**.
- Go to the **Send** tab in the wallet.
- You shall notice, any new Send transaction created will have the preferred setting as chosen by the user.<br> (Try clicking on Add recipient or even restarting the Node in GUI)

Attaching ScreenRecordings to explain the added feature. 

> Master.mov: Master Branch

https://user-images.githubusercontent.com/54016434/127763378-be91837d-d0ab-4ae5-87c0-d303fa70a336.mov

> PR.mov: PullRequest

https://user-images.githubusercontent.com/54016434/127763404-05b834c1-4082-4fbd-9b05-1528ac898a21.mov


Close #386

**Review by ShaMan239 — COMMENTED** [review]
tACK 4dc91813aca000336cec2cb40244d006a88b7d6a
Tested on Ubuntu 20.04

This PR is adding a feature in the wallet settings so that the GUI will remember the default behavior of whether to subtract the fee from the transaction or not. As said by op this setting will be remembered even when the GUI is closed and reopened. This is done by setting a bool value based on if the checkbox is selected in settings or not, and then using that bool value to determine the default state of the SubFeeFromAmount option.
The PR is doing what it describes in the description. It is also not modifying any part of the code, which it should be modifying.
Adding some screenshots to show the proper working of the PR. The PR was compiled and ran on the signet chain.
1. The subtract fee from amount option in settings (highlighted in green)
Screenshot from 2021-08-02 19-54-29

    

2. Selection of “subtract fee from amount” option by default:
Screenshot from 2021-08-02 19-54-41

In my opinion, this PR is an improvement over the current master. For those who often use the “Subtract fee from amount” option, this feature can be quite handy.

File: src/qt/forms/optionsdialog.ui [hunk 1]
```diff
@@ -206,6 +206,16 @@
           <string>Expert</string>
          </property>
          <layout class="QVBoxLayout" name="verticalLayout_2">
+          <item>
+           <widget class="QCheckBox" name="subFeeFromAmount">
+            <property name="toolTip">
+             <string>Whether to Subtact Fee From Amount or not.</string>
+            </property>
+            <property name="text">
+             <string>Subtract &amp;Fee From Amount</string>
```
Comment (jarolrod) [inline src/qt/forms/optionsdialog.ui:215]: For any string in a UI file, you can add an `extracomment` field within the opening string tag. The value of this field is equivalent to a translator comment in our qt code

```suggestion
             <string extracomment="An options menu setting to always subtract the fee from the amount a user is sending.">Subtract &amp;Fee From Amount</string>
```

File: src/qt/forms/optionsdialog.ui [hunk 2]
```diff
@@ -206,6 +206,16 @@
           <string>Expert</string>
          </property>
          <layout class="QVBoxLayout" name="verticalLayout_2">
+          <item>
+           <widget class="QCheckBox" name="subFeeFromAmount">
+            <property name="toolTip">
+             <string>Whether to Subtact Fee From Amount or not.</string>
```
Comment (jarolrod) [inline src/qt/forms/optionsdialog.ui:212]:
```suggestion
             <string>Whether to always subtract fee from amount or not.</string>
```

**Review by jarolrod — COMMENTED** [review]
Concept ACK, please add translator comments

**Comment by psancheti110** [comment]
PR Updated from 4dc91813aca000336cec2cb40244d006a88b7d6a -> 17d7b21a862835a768fd436dc43da1bb60b4fa25
Addressed changes suggested by @jarolrod ([PR#390 Comment](https://github.com/bitcoin-core/gui/pull/390#pullrequestreview-720718995))

File: src/qt/forms/optionsdialog.ui [hunk 3]
```diff
@@ -206,6 +206,16 @@
           <string>Expert</string>
          </property>
          <layout class="QVBoxLayout" name="verticalLayout_2">
+          <item>
+           <widget class="QCheckBox" name="subFeeFromAmount">
+            <property name="toolTip">
+             <string extracomment="This is a toolTip, Located at Options dialog under the Wallet tab. Select default setting for subtract the fee from the amount checkbox while creating new recipients.">Default setting to whether Subtact Fee From Amount or not.</string>
+            </property>
+            <property name="text">
+             <string extracomment="Located at Options dialog under the Wallet tab. Select default setting for subtract the fee from the amount checkbox while creating new recipients.">Subtract &amp;Fee From Amount</string>
```
Comment (jarolrod) [inline src/qt/forms/optionsdialog.ui:215]:
```suggestion
             <string extracomment="An options menu setting to always subtract the fee from the amount a user is sending.">Subtract &amp;Fee From Amount</string>
```

File: src/qt/forms/optionsdialog.ui [hunk 4]
```diff
@@ -206,6 +206,16 @@
           <string>Expert</string>
          </property>
          <layout class="QVBoxLayout" name="verticalLayout_2">
+          <item>
+           <widget class="QCheckBox" name="subFeeFromAmount">
+            <property name="toolTip">
+             <string extracomment="This is a toolTip, Located at Options dialog under the Wallet tab. Select default setting for subtract the fee from the amount checkbox while creating new recipients.">Default setting to whether Subtact Fee From Amount or not.</string>
```
Comment (jarolrod) [inline src/qt/forms/optionsdialog.ui:212]:
```suggestion
             <string extracomment="Tooltip text for options window setting that sets subtracting the fee from a sending amount as default">Whether to always subtract fee from amount or not.</string>
```

**Review by jarolrod — COMMENTED** [review]
Concept ACK

Existing recipients should respond to the 'SubFeeFromAmountChanged' signal.

There are scenarios where you have multiple recipients filled out. After you've filled them out, you now want all of them to 'Subtract fee from amount'. If we allow the existing recipients to respond to the 'SubFeeFromAmountChanged' signal, then you don't have to manually check every box. You can go into settings, set the option, come back and all of your recipients will not have the 'Subtract Fee from amount' box checked.

Similarly, if you already had the option set and created a bunch of recipients; you could unselect the option in settings and then all the boxes become unchecked.

**Comment by hebasto** [comment]
@jarolrod 
> Existing recipients should respond to the 'SubFeeFromAmountChanged' signal.

I don't agree.

The Options Dialog sets the _default_ value of the "Subtract Fee..." setting. That means it is used only once per every recipient -- just initialize its own checkbox. In other words, it is not supposed to be a batch switcher.

> There are scenarios where you have multiple recipients filled out. After you've filled them out, you now want all of them to 'Subtract fee from amount'. If we allow the existing recipients to respond to the 'SubFeeFromAmountChanged' signal, then you don't have to manually check every box. You can go into settings, set the option, come back and all of your recipients will not have the 'Subtract Fee from amount' box checked.

I'm not sure if this scenario would be ever convenient for users.

File: src/qt/sendcoinsentry.cpp [hunk 5]
```diff
@@ -97,7 +97,9 @@ void SendCoinsEntry::clear()
     ui->payTo->clear();
     ui->addAsLabel->clear();
     ui->payAmount->clear();
-    ui->checkboxSubtractFeeFromAmount->setCheckState(Qt::Unchecked);
+    if(model && model->getOptionsModel()){
```
Comment (hebasto) [inline src/qt/sendcoinsentry.cpp:100]:
```suggestion
    if (model && model->getOptionsModel()) {
```

There is a very useful tool to check commit code style before pushing -- [`clang-format-diff.py`](https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy).

File: src/qt/optionsmodel.h [hunk 6]
```diff
@@ -61,6 +61,7 @@ class OptionsModel : public QAbstractListModel
         Language,               // QString
         UseEmbeddedMonospacedFont, // bool
         CoinControlFeatures,    // bool
+        SubFeeFromAmount,       //bool
```
Comment (hebasto) [inline src/qt/optionsmodel.h:64]: nit, why not follow the surrounding style?
```suggestion
        SubFeeFromAmount,       // bool
```

**Comment by psancheti110** [comment]
PR Updated from 17d7b21a862835a768fd436dc43da1bb60b4fa25 -> 50216bc8e28415ce9615b7ed1108c6730280d31b . Click [_here_](https://github.com/bitcoin-core/gui/compare/17d7b21a862835a768fd436dc43da1bb60b4fa25..50216bc8e28415ce9615b7ed1108c6730280d31b) to check diff.

Addressed changes suggested by @hebasto about coding style using `clang-format-diff.py`

File: src/qt/forms/optionsdialog.ui (hunk 3 above)
Comment (jarolrod) [reply src/qt/forms/optionsdialog.ui:215]: @psancheti110 updated strings to remove the word always as it is not appropriate

File: src/qt/forms/optionsdialog.ui (hunk 4 above)
Comment (jarolrod) [reply src/qt/forms/optionsdialog.ui:212]: updated suggestion, see https://github.com/bitcoin-core/gui/pull/390#discussion_r684421703

**Comment by jarolrod** [comment]
Approach ACK

I rescind my comments here about the behavior of this option. The appropriate place for a batch switch would be in the Send tab itself. Please update the translator comments 🥃

**Comment by hebasto** [comment]
Why a new "subFeeFromAmount" checkbox is placed within the "Expert" group? Shouldn't it be an independent option?

**Comment by psancheti110** [comment]
@hebasto 

> Why a new "subFeeFromAmount" checkbox is placed within the "Expert" group? Shouldn't it be an independent option?


Agreed, It doesn't belong to an "expert" group. That's quite some observation there. 

Will make the required changes and add.

**Comment by psancheti110** [comment]
PR Updated from 50216bc8e28415ce9615b7ed1108c6730280d31b -> https://github.com/bitcoin-core/gui/commit/fb194c347e95e20bb682de75e51db084c7a4855c. Click [_here_](https://github.com/bitcoin-core/gui/compare/50216bc8e28415ce9615b7ed1108c6730280d31b..fb194c347e95e20bb682de75e51db084c7a4855c) to check diff.

Addressed changes suggested by @hebasto for moving SubFeeFromAmount option outside "Expert" group and translator comments as suggested by @jarolrod.

**Review by hebasto — COMMENTED** [review]
Approach ACK fb194c347e95e20bb682de75e51db084c7a4855c, going to test and review more closely.

File: src/qt/forms/optionsdialog.ui [hunk 7]
```diff
@@ -200,6 +200,16 @@
        <string>W&amp;allet</string>
       </attribute>
       <layout class="QVBoxLayout" name="verticalLayout_Wallet">
+      <item>
+       <widget class="QCheckBox" name="subFeeFromAmount">
+        <property name="toolTip">
+         <string extracomment="Tooltip text for options window setting that sets subtracting the fee from a sending amount as default.">Whether to set subtract fee from amount as default or not.</string>
+        </property>
+        <property name="text">
+         <string extracomment="An options menu setting to set subtracting the fee from the amount a user is sending as default.">Subtract &amp;Fee From Amount</string>
```
Comment (hebasto) [inline src/qt/forms/optionsdialog.ui:209]:
```suggestion
         <string extracomment="An Options window setting to set subtracting the fee from a sending amount as default.">Subtract &amp;fee from amount by default</string>
```

**Review by hebasto — COMMENTED** [review]
Tested fb194c347e95e20bb682de75e51db084c7a4855c on Linux Mint 20.2 (Qt 5.12.8), and it works flawlessly, and it looks great:

Screenshot from 2021-08-08 22-54-14

**Comment by psancheti110** [comment]
PR Updated from fb194c347e95e20bb682de75e51db084c7a4855c -> 54ce50987092b60a77677a4e59e8fd0a250b7bcd. Click [_here_](https://github.com/bitcoin-core/gui/compare/fb194c347e95e20bb682de75e51db084c7a4855c..54ce50987092b60a77677a4e59e8fd0a250b7bcd) to check diff.

Addressed changes suggested by @hebasto. Moved the definitions of SubFeeFromAmount under `#ifdef ENABLE_WALLET` because this functionality cannot work without a wallet being loaded.

File: src/qt/forms/optionsdialog.ui [hunk 8]
```diff
@@ -51,20 +51,20 @@
...
-          </property>
-          <property name="text">
-           <string>Prune &amp;block storage to</string>
-          </property>
-         </widget>
-        </item>
-        <item>
-         <widget class="QSpinBox" name="pruneSize"/>
-        </item>
+        <layout class="QHBoxLayout" name="horizontalLayout_Main_Prune">
```
Comment (jarolrod) [inline src/qt/forms/optionsdialog.ui:54]: nit, fixing the indentations could have gone in its own commit

File: src/qt/forms/optionsdialog.ui [hunk 9]
```diff
@@ -200,6 +200,16 @@
        <string>W&amp;allet</string>
       </attribute>
       <layout class="QVBoxLayout" name="verticalLayout_Wallet">
+       <item>
+        <widget class="QCheckBox" name="subFeeFromAmount">
+         <property name="toolTip">
+          <string extracomment="Tooltip text for Options window setting that sets subtracting the fee from a sending amount as default.">Whether to set subtract fee from amount as default.</string>
```
Comment (jarolrod) [inline src/qt/forms/optionsdialog.ui:206]:
```suggestion
          <string extracomment="Tooltip text for Options window setting that sets subtracting the fee from a sending amount as default.">Whether to set subtract fee from amount as default or not.</string>
```

**Review by jarolrod — COMMENTED** [review]
Almost there, one last thing: https://github.com/bitcoin-core/gui/pull/390#discussion_r685659408

**Comment by psancheti110** [comment]
PR Updated from 54ce50987092b60a77677a4e59e8fd0a250b7bcd -> ec25093a5793451aa997b5ff8a05d1631b41e05c. Click [_here_](https://github.com/bitcoin-core/gui/compare/54ce50987092b60a77677a4e59e8fd0a250b7bcd..ec25093a5793451aa997b5ff8a05d1631b41e05c) to check diff.

**Review by hebasto — APPROVED** [review]
ACK 7bf1a2bf569fe9356592b94531c0bbe0093ca302, tested on Linux Mint 20.2 (Qt 5.12.8).

Also I agree with the _"qt, refactor: Fix indentation"_ commit (7bf1a2bf569fe9356592b94531c0bbe0093ca302) because every time `optionsdialog.ui` being opened in the Qt Designer, the latter tries to fix indentation, and produces unexpected changes.

**Comment by jarolrod** [comment]
tACK 7bf1a2bf569fe9356592b94531c0bbe0093ca302

tested on macOS 12 and Ubuntu 20.04. The setting works nicely and is a definite UX improvement. Nice job 🥃

**Review by ShaMan239 — COMMENTED** [review]
tACK 7bf1a2bf569fe9356592b94531c0bbe0093ca302
Tested on Ubuntu 20.04

Changes that OP made since my previous review:
1. Formatting is fixed according to clang-format.
2. Moved case SubFeeAmount under #ifdef ENABLE_WALLET to prevent this feature from working when wallet functionality is disabled.
3. The AddSubFeeAmount options have been moved out of the Expert menu.
4. Indentations in optionsdialog.ui have been fixed not to create unnecessary indentation changes when the file is opened in QTDesigner.

Screenshot:
Screenshot from 2021-08-10 20-13-32

Just want to add one small thing; you should modify the video describing the PR, as it is still showing the option under the Expert menu :)
<p>

File: src/qt/optionsmodel.h [hunk 10]
```diff
@@ -123,6 +126,7 @@ class OptionsModel : public QAbstractListModel
 Q_SIGNALS:
     void displayUnitChanged(int unit);
     void coinControlFeaturesChanged(bool);
+    void SubFeeFromAmountChanged(bool);
```
Comment (meshcollider) [inline src/qt/optionsmodel.h:129]: Why is this Q_SIGNAL needed if it isn't `connect()`ed anywhere?

Comment (psancheti110) [reply src/qt/optionsmodel.h]: Addressed the suggested changes in the recent commit.

**Comment by psancheti110** [comment]
PR Updated from 7bf1a2bf569fe9356592b94531c0bbe0093ca302 -> 62b125fd197879f112322a1f67a318d6ab22e67a. Click [_here_](https://github.com/bitcoin-core/gui/compare/7bf1a2bf569fe9356592b94531c0bbe0093ca302..62b125fd197879f112322a1f67a318d6ab22e67a) to check diff.

Addressed changes suggested by @meshcollider.

**Review by hebasto — APPROVED** [review]
re-ACK 62b125fd197879f112322a1f67a318d6ab22e67a, only removed the unused `SubFeeFromAmountChanged` signal since my [previous](https://github.com/bitcoin-core/gui/pull/390#pullrequestreview-726531766) review.

**Comment by Talkless** [comment]
tACK 62b125fd197879f112322a1f67a318d6ab22e67a, tested on Debian Sid with 5.15.2 and it works as described.

**Review by meshcollider — COMMENTED** [review]
utACK 62b125fd197879f112322a1f67a318d6ab22e67a

**Comment by psancheti110** [comment]
**_A big thanks to all the contributors who took out time to test and review my PR. 
Appreciate each and every comment made here by fellow mates_**  🙌🏻

**Pull request merged**